baremetal/metal3machine_manager.go

```diff
@@ -552,6 +580,67 @@ func (m *MachineManager) Delete(ctx context.Context) error {
 			return &RequeueAfterError{RequeueAfter: requeueAfter}
 		}
 
+		// Fetch corresponding Metal3MachineTemplate, to see if NodeReuse
+		// feature is enabled. If set to true, check the machine role. In case
+		// machine role is ControlPlane, set nodeReuseLabelName to KubeadmControlPlane
+		// name, otherwise to MachineDeployment name.
+		m.Log.Info("Getting Metal3MachineTemplate")
+		m3mt := &capm3.Metal3MachineTemplate{}
+		if m.Metal3Machine == nil {
+			return errors.New("Metal3Machine associated with Metal3MachineTemplate is not found")
+		}
+		if m.hasTemplateAnnotation() {
+			m3mtKey := client.ObjectKey{
+				Name:      m.Metal3Machine.ObjectMeta.GetAnnotations()[capi.TemplateClonedFromNameAnnotation],
+				Namespace: m.Metal3Machine.Namespace,
+			}
+			if err := m.client.Get(ctx, m3mtKey, m3mt); err != nil {
+				// we are here, because while normal deprovisioning, Metal3MachineTemplate will be deleted first
+				// and we can't get it even though Metal3Machine has reference to it. We consider it nil and move
+				// forward with normal deprovisioning.
+				m3mt = nil
+				m.Log.Info("Metal3MachineTemplate associated with Metal3Machine is deleted")
+			} else {
+				// in case of upgrading, Metal3MachineTemplate is still and not deleted there and we can get it,
+				// so we can check for node reuse feature in the next step.
+				m.Log.Info("Found Metal3machineTemplate", "metal3machinetemplate", m3mtKey.Name)
+			}
+		}
+		if m3mt != nil {
+			if m3mt.Spec.NodeReuse {
+				// Check if machine is ControlPlane
+				if m.isControlPlane() {
+					// Fetch KubeadmControlPlane name for controlplane machine
+					m.Log.Info("Fetch KubeadmControlPlane name")
+					kcpName, err := m.getKubeadmControlPlaneName(ctx)
+					if err != nil {
+						return err
+					}
+					m.Log.Info("Fetched KubeadmControlPlane name:", "kubeadmcontrolplane", kcpName)
+					if host.Labels == nil {
```

Comment:
shall we do these somewhere around line 611 instead of doing it both in `if` and `else` to reduce code entropy ?

Reply:
Fixed.